components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/SAMLSSOServiceProviderManagerTest.java

```diff
@@ -0,0 +1,356 @@
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider2, TENANT_ID);
+
+        SAMLSSOServiceProviderDO[] serviceProviders = samlSSOServiceProviderManager.getServiceProviders(TENANT_ID);
+
+        assertEquals(serviceProviders, new SAMLSSOServiceProviderDO[]{sampleServiceProvider1, sampleServiceProvider2});
+    }
+
+    @Test
+    public void testGetServiceProvidersWithDataAccessException() throws Exception {
+
+        try (MockedStatic<JdbcUtils> jdbcUtilsMockedStatic = mockStatic(JdbcUtils.class)) {
+            NamedJdbcTemplate namedJdbcTemplate = mock(NamedJdbcTemplate.class);
+            jdbcUtilsMockedStatic.when(JdbcUtils::getNewNamedJdbcTemplate).thenReturn(namedJdbcTemplate);
+            doThrow(new DataAccessException("Data access error")).when(namedJdbcTemplate)
+                    .executeQuery(any(), any(), any());
+
+            assertThrows(IdentityException.class, () -> samlSSOServiceProviderManager.getServiceProviders(TENANT_ID));
+        }
+    }
+
+    @Test
+    public void testGetServiceProvider() throws Exception {
+
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID);
+
+        SAMLSSOServiceProviderDO serviceProviderFromStorage =
+                samlSSOServiceProviderManager.getServiceProvider(getIssuerWithQualifier(ISSUER1), TENANT_ID);
+
+        assertEquals(serviceProviderFromStorage, sampleServiceProvider1);
+    }
+
+    @Test
+    public void testProcessGetServiceProviderWithDataAccessException() throws Exception {
+
+        try (MockedStatic<JdbcUtils> jdbcUtilsMockedStatic = mockStatic(JdbcUtils.class)) {
+            NamedJdbcTemplate namedJdbcTemplate = mock(NamedJdbcTemplate.class);
+            jdbcUtilsMockedStatic.when(JdbcUtils::getNewNamedJdbcTemplate).thenReturn(namedJdbcTemplate);
+            doThrow(new DataAccessException("Data access error")).when(namedJdbcTemplate)
+                    .fetchSingleRecord(any(), any(), any());
+
+            assertThrows(IdentityException.class,
+                    () -> samlSSOServiceProviderManager.getServiceProvider(ISSUER1, TENANT_ID));
+        }
+    }
+
+    @Test
+    public void testIsServiceProviderExists() throws Exception {
+
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID);
+        assertTrue(samlSSOServiceProviderManager.isServiceProviderExists(getIssuerWithQualifier(ISSUER1), TENANT_ID));
+    }
+
+    @Test
+    public void testNonExistingIsServiceProviderExists() throws Exception {
+
+        assertFalse(samlSSOServiceProviderManager.isServiceProviderExists(getIssuerWithQualifier(ISSUER1), TENANT_ID));
+    }
+
+    @Test
+    public void testRemoveServiceProvider() throws Exception {
+
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID);
+        samlSSOServiceProviderManager.removeServiceProvider(getIssuerWithQualifier(ISSUER1), TENANT_ID);
```

Comment:
Better to assert the result of `removeServiceProvider` as well in here.

Reply:
Resolved in [Introduce JDBC based persistence for SAML](https://github.com/wso2/carbon-identity-framework/pull/6241/commits/d1e67ce649f7a5d2e7e62f364ce93815f347e7b4)